File: app/src/main/java/com/benhero/glstudio/base/GLScaleAnimation.java
package com.benhero.glstudio.base;

import android.opengl.Matrix;

/**
 * GL缩放动画
 *
 * @author chenbenbin
 */
public class GLScaleAnimation extends GLAnimation {
    private float mFromX = 1.0f;
    private float mToX = 1.0f;
    private float mFromY = 1.0f;
    private float mToY = 1.0f;
    private int mPivotXType;

    public float getFromX() {
        return mFromX;
    }

    public void setFromX(float fromX) {
        mFromX = fromX;
    }

    public float getToX() {
        return mToX;
    }

    public void setToX(float toX) {
        mToX = toX;
    }

    public float getFromY() {
        return mFromY;
    }

    public void setFromY(float fromY) {
        mFromY = fromY;
    }

    public float getToY() {
        return mToY;
    }

    public void setToY(float toY) {
        mToY = toY;
    }

    @Override
    protected void applyTransformation(float percent, GLObject object) {
        super.applyTransformation(percent, object);
        float currentX = (mToX - mFromX) * percent + mFromX;
        float currentY = (mToY - mFromY) * percent + mFromY;
        Matrix.scaleM(object.getPositionMatrix(), 0,
                currentX, currentY, 1);
    }
}

File: app/src/main/java/com/benhero/glstudio/base/GLScaleAnimation.kt
package com.benhero.glstudio.base

import android.opengl.Matrix

/**
 * GL缩放动画
 *
 * @author chenbenbin
 */
class GLScaleAnimation : GLAnimation() {
    var fromX = 1.0f
    var toX = 1.0f
    var fromY = 1.0f
    var toY = 1.0f
    private val mPivotXType: Int = 0

    override fun applyTransformation(percent: Float, `object`: GLObject) {
        super.applyTransformation(percent, `object`)
        val currentX = (toX - fromX) * percent + fromX
        val currentY = (toY - fromY) * percent + fromY
        Matrix.scaleM(`object`.positionMatrix, 0,
                currentX, currentY, 1f)
    }
}
